compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -49,10 +49,23 @@ impl Ssa {
             function.dfg.set_function_purities(purities.clone());
         }
 
+        purity_analysis_post_check(&self);
+
         self
     }
 }
 
+/// Post-check condition for [Function::remove_bit_shifts].
+///
+/// Returns `true` if:
+///   - all functions have a purity status attached to it.
+///
+/// Otherwise returns `false`.
+fn purity_analysis_post_check(ssa: &Ssa) -> bool {
+    // TODO: We have N different copies of the purity hashmap (one per function), ideally this would be deduplicated.
```

Comment:
Should an issue be captured for this?

Reply:
`FunctionPurities` are contained within an Arc which provides shared ownership.